pkg/utils/tasks/tasks.go

```diff
@@ -31,64 +35,233 @@ type TaskManager struct {
+	h.attempts++
+	h.err.Append(err)
+
+	// Check if we have retries left
+	if h.attempts <= h.opts.retries {
+		// Schedule the retry be added back to the queue
+		h.retryAt = time.Now().Add(h.opts.backoff.NextBackOff())
+		log.Warn().
+			Err(err).
+			Time("retry_at", h.retryAt).
+			Int("attempts", h.attempts).
+			Int("retries_remaining", h.opts.retries-h.attempts).
+			Msg("async task failed, retrying")
+
+		h.parent.queueTask(h)
+		return
+	}
+
+	// At this point we've exhausted all possible retries, so log the error.
+	h.err.Since(h.scheduled)
+	h.err.Log(log.Logger)
+	h.err.Capture(sentry.GetHubFromContext(h.ctx))
+}
+
+// TaskScheduler runs as a separate Go routine, listening for tasks on the retry
+// channel and queueing them for a worker when their backoff period has expired.
+func TaskScheduler(wg *sync.WaitGroup, queue <-chan *TaskHandler, tasks chan<- *TaskHandler, stop <-chan struct{}, interval time.Duration) {
 	defer wg.Done()
-	for handler := range queue {
-		handler.task.Do(handler.ctx)
+
+	// Hold tasks awaiting retry and queue them every tick if ready
+	// TODO: how do we test to ensure there is no memory leak?
+	pending := make([]*TaskHandler, 0, 64)
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case task := <-queue:
+			// Check if the task is a retry task that needs to be held
+			if !task.retryAt.IsZero() && time.Now().Before(task.retryAt) {
+				pending = append(pending, task)
+				continue
+			}
+
+			// Otherwise send the task to the worker queue immediately
+			// Do not block the send; if no workers are available, append the task to
+			// the pending data structure. This will reduce the backpressure on the
+			// queue but also prevent deadlocks where there are more retries than
+			// workers available and no one can make progress.
+			select {
+			case tasks <- task:
+				continue
+			default:
+				pending = append(pending, task)
+			}
+
+		case now := <-ticker.C:
+			// Check all of the pending tasks to see if any are ready to be queued
+			for i, task := range pending {
+				if task.retryAt.IsZero() || task.retryAt.Before(now) {
+					// The task is ready to retry; queue it up and delete it from pending
+					// Note: this is a non-blocking write to tasks in case there are no
+					// workers available to handle the current task.
+					select {
+					case tasks <- task:
+						pending[i] = nil
+					default:
+						continue
+					}
+				}
+			}
+
+			// Prevent memory leaks by shifting tasks to deleted spots without allocation
+			i := 0
+			for _, task := range pending {
+				if task != nil {
+					pending[i] = task
+					i++
+				}
+			}
+
+			// Compute the new capacity, shrinking it if necessary to prevent leaks.
+			newcap := cap(pending)
+			if i+64 < newcap {
+				newcap = i + 64
+			}
+
+			pending = pending[:i:newcap]
+			log.Trace().Int("pending_length", len(pending)).Int("pending_capacity", cap(pending)).Msg("async task scheduler memory usage")
```

Comment:
You didn't have this problem in your code since you we're copying into a `remaining` slice; but copying is a fairly heavy weight operation so this code prevents the memory leak while reducing the number of allocations and does not perform a copy. I think this is important for code that would be running once a second.

Reply:
On monday, I might have you do some profiling to make sure that we've prevented the memory leak - would you be interested? If so, could you create a story for us and a meeting to take on that task?

Reply:
I put a meeting on the calendar, I could also do this morning if that time doesn't work.